### PR TITLE
feat(line): draw charts related to line geometry including parallel coordinate.

### PR DESCRIPTION
**Pull request by pearmini**
Draw charts related to line geometry including parallel coordinate. (Test coverage: 96%)

 **Features**

- Runtime
  - Add position channel for parallel coordinate.
  - Add size channel.
  - Optimize scale inference: Identity scale for non-primitive value.
  - Optimize component inference: Multiple axes in the center.
  - Sort marks and components by zIndex.
  - Support inferring statistic with infer components.
- Geometry: Line
- Shape: Line, Smooth
- Scale: Time, Point
- Infer: MaybeStackY, MaybeSeries, MaybeSplitPosition
- Statistic: StackY, SplitPosition
- Coordinate: Parallel
- Transform: FilterBy

**Bug Fixes**
- Components placed in wrong direction. They should be placed from center to border.
- Guide options is not valid.

**Demo**

Basic line chart.

image

```js
const options = {
  type: 'line',
  data: [
    { year: '1991', value: 15468 },
    { year: '1992', value: 16100 },
    { year: '1993', value: 15900 },
    { year: '1994', value: 17409 },
    { year: '1995', value: 17000 },
    { year: '1996', value: 31056 },
    { year: '1997', value: 31982 },
    { year: '1998', value: 32040 },
    { year: '1999', value: 33233 },
  ],
  encode: {
    x: 'year',
    y: 'value',
  },
}
```

Smooth line chart.

image

```js
const options = {
  type: 'line',
  data: [
    { year: '1991', value: 15468 },
    { year: '1992', value: 16100 },
    { year: '1993', value: 15900 },
    { year: '1994', value: 17409 },
    { year: '1995', value: 17000 },
    { year: '1996', value: 31056 },
    { year: '1997', value: 31982 },
    { year: '1998', value: 32040 },
    { year: '1999', value: 33233 },
  ],
  encode: {
    x: 'year',
    y: 'value',
    shape: 'smooth'
  },
}
```
Alpha for smooth shape.

image

```js
const options = {
  type: 'line',
  data: [
    { year: '1991', value: 15468 },
    { year: '1992', value: 16100 },
    { year: '1993', value: 15900 },
    { year: '1994', value: 17409 },
    { year: '1995', value: 17000 },
    { year: '1996', value: 31056 },
    { year: '1997', value: 31982 },
    { year: '1998', value: 32040 },
    { year: '1999', value: 33233 },
  ],
  transform: [
    {
      type: ({ alphas = [] }) => {
        if (!Array.isArray(alphas)) return (data) => data;
        return (data) =>
          alphas.flatMap((alpha) =>
            data.map((d) => ({
              ...d,
              alpha,
            })),
          );
      },
      alphas: [0, 0.25, 0.5, 0.75, 1],
    },
  ],
  scale: {
    color: { field: 'alpha' },
  },
  encode: {
    x: 'year',
    y: 'value',
    color: (d) => `${d.alpha}`,
    shape: {
      type: 'transform',
      value: (d) => ({
        type: 'smooth',
        alpha: d.alpha,
      }),
    },
  },
}
```
Multiple series of line.

image

```js
const options = {
  type: 'line',
  data: [
    { month: 'Jan', city: 'Tokyo', temperature: 7 },
    { month: 'Jan', city: 'London', temperature: 3.9 },
    { month: 'Feb', city: 'Tokyo', temperature: 6.9 },
    { month: 'Feb', city: 'London', temperature: 4.2 },
    { month: 'Mar', city: 'Tokyo', temperature: 9.5 },
    { month: 'Mar', city: 'London', temperature: 5.7 },
    { month: 'Apr', city: 'Tokyo', temperature: 14.5 },
    { month: 'Apr', city: 'London', temperature: 8.5 },
    { month: 'May', city: 'Tokyo', temperature: 18.4 },
    { month: 'May', city: 'London', temperature: 11.9 },
    { month: 'Jun', city: 'Tokyo', temperature: 21.5 },
    { month: 'Jun', city: 'London', temperature: 15.2 },
    { month: 'Jul', city: 'Tokyo', temperature: 25.2 },
    { month: 'Jul', city: 'London', temperature: 17 },
    { month: 'Aug', city: 'Tokyo', temperature: 26.5 },
    { month: 'Aug', city: 'London', temperature: 16.6 },
    { month: 'Sep', city: 'Tokyo', temperature: 23.3 },
    { month: 'Sep', city: 'London', temperature: 14.2 },
    { month: 'Oct', city: 'Tokyo', temperature: 18.3 },
    { month: 'Oct', city: 'London', temperature: 10.3 },
    { month: 'Nov', city: 'Tokyo', temperature: 13.9 },
    { month: 'Nov', city: 'London', temperature: 6.6 },
    { month: 'Dec', city: 'Tokyo', temperature: 9.6 },
    { month: 'Dec', city: 'London', temperature: 4.8 },
  ],
  encode: {
    x: 'month',
    y: 'temperature',
    color: 'city',
  },
};
```

Time scale for x channel.

image

```js
const options = {
  type: 'line',
  transform: [
    {
      type: 'fetch',
      url: 'https://gw.alipayobjects.com/os/bmw-prod/ab55d10f-24da-465a-9eba-87ac4b7a83ec.json',
    },
  ],
  encode: {
    // Convert to Date Object will infer time scale for this channel.
    x: (d) => new Date(d.Date), 
    y: 'Close',
  },
};
```

Size channel for line.

image

```js
const options = {
      type: 'line',
      data: [
        { month: 'Jan', city: 'Tokyo', temperature: 7 },
        { month: 'Jan', city: 'London', temperature: 3.9 },
        { month: 'Feb', city: 'Tokyo', temperature: 6.9 },
        { month: 'Feb', city: 'London', temperature: 4.2 },
        { month: 'Mar', city: 'Tokyo', temperature: 9.5 },
        { month: 'Mar', city: 'London', temperature: 5.7 },
        { month: 'Apr', city: 'Tokyo', temperature: 14.5 },
        { month: 'Apr', city: 'London', temperature: 8.5 },
        { month: 'May', city: 'Tokyo', temperature: 18.4 },
        { month: 'May', city: 'London', temperature: 11.9 },
        { month: 'Jun', city: 'Tokyo', temperature: 21.5 },
        { month: 'Jun', city: 'London', temperature: 15.2 },
        { month: 'Jul', city: 'Tokyo', temperature: 25.2 },
        { month: 'Jul', city: 'London', temperature: 17 },
        { month: 'Aug', city: 'Tokyo', temperature: 26.5 },
        { month: 'Aug', city: 'London', temperature: 16.6 },
        { month: 'Sep', city: 'Tokyo', temperature: 23.3 },
        { month: 'Sep', city: 'London', temperature: 14.2 },
        { month: 'Oct', city: 'Tokyo', temperature: 18.3 },
        { month: 'Oct', city: 'London', temperature: 10.3 },
        { month: 'Nov', city: 'Tokyo', temperature: 13.9 },
        { month: 'Nov', city: 'London', temperature: 6.6 },
        { month: 'Dec', city: 'Tokyo', temperature: 9.6 },
        { month: 'Dec', city: 'London', temperature: 4.8 },
      ],
      encode: {
        x: 'month',
        y: 'temperature',
        color: 'city',
        size: 'city', // Size channel
      },
    }
```

Parallel coordinate.

image

```js
const options = {
  type: 'line',
  width: 720,
  paddingLeft: 80,
  transform: [
    {
      type: 'fetch',
      url: 'https://gw.alipayobjects.com/os/bmw-prod/96cd81b5-54a4-4fe8-b778-502b2114df58.json',
      callback: ({ Year, ...rest }) => ({
        Year: new Date(Year),
        ...rest,
      }),
    },
    {
      type: 'filterBy',
       // Filter data with defined Horsepower and Miles_per_Gallon.
      fields: ['Horsepower', 'Miles_per_Gallon'],
    },
  ],
  coordinate: [{ type: 'parallel' }],
  scale: {
    // zIndex of mark is default to 0.
    // zIndex of component is default to -1.
    // Set zIndex to 1 for component to draw above marks.
    'position[1]': { nice: true, guide: { zIndex: 1 } }, 
    'position[2]': { nice: true, guide: { zIndex: 1 } },
    'position[3]': { nice: true, guide: { zIndex: 1 } },
    'position[4]': { nice: true, guide: { zIndex: 1 } },
    'position[5]': { nice: true, guide: { zIndex: 1 } },
    'position[6]': { nice: true, guide: { zIndex: 1 } },
  },
  encode: {
    position: [
      'Cylinders',
      'Displacement',
      'Weight_in_lbs',
      'Horsepower',
      'Acceleration',
      'Miles_per_Gallon',
      'Year',
    ],
    color: 'Origin',
    size: 1.01,
  },
  style: {
    strokeOpacity: 0.3,
  },
};
```
Parallel coordinate with animation.

ParallelAnimation

```js
const options = {
  type: 'line',
  width: 720,
  paddingLeft: 80,
  transform: [
    {
      type: 'fetch',
      url: 'https://gw.alipayobjects.com/os/bmw-prod/96cd81b5-54a4-4fe8-b778-502b2114df58.json',
      callback: ({ Year, ...rest }) => ({
        Year: new Date(Year),
        ...rest,
      }),
    },
    {
      type: 'filterBy',
      fields: ['Horsepower', 'Miles_per_Gallon'],
    },
  ],
  coordinate: [{ type: 'parallel' }],
  scale: {
    'position[1]': { nice: true, guide: { zIndex: 1 } },
    'position[2]': { nice: true, guide: { zIndex: 1 } },
    'position[3]': { nice: true, guide: { zIndex: 1 } },
    'position[4]': { nice: true, guide: { zIndex: 1 } },
    'position[5]': { nice: true, guide: { zIndex: 1 } },
    'position[6]': { nice: true, guide: { zIndex: 1 } },
  },
  statistic: [{ type: 'stackEnter', by: ['color'] }],
  encode: {
    position: [
      'Cylinders',
      'Displacement',
      'Weight_in_lbs',
      'Horsepower',
      'Acceleration',
      'Miles_per_Gallon',
      'Year',
    ],
    color: 'Cylinders',
    size: 1.01,
    enterDuration: 3000,
  },
  style: {
    strokeOpacity: 0.3,
  },
}
```

Radar chart.

image

```js
const options = {
  type: 'line',
  width: 720,
  paddingLeft: 80,
  transform: [
    {
      type: 'fetch',
      url: 'https://gw.alipayobjects.com/os/bmw-prod/96cd81b5-54a4-4fe8-b778-502b2114df58.json',
      callback: ({ Year, ...rest }) => ({
        Year: new Date(Year),
        ...rest,
      }),
    },
    {
      type: 'filterBy',
      fields: ['Horsepower', 'Miles_per_Gallon'],
    },
  ],
  coordinate: [{ type: 'parallel' }, {type:'polar'}],
  encode: {
    position: [
      'Cylinders',
      'Displacement',
      'Weight_in_lbs',
      'Horsepower',
      'Acceleration',
      'Miles_per_Gallon',
      'Year',
      'Cylinders' // Close the line.
    ],
    color: 'Origin',
    size: 1.01,
  },
  style: {
    strokeOpacity: 0.3,
  },
}
```

### CODE REVIEW

**Comment by visiky** [comment]
从 PR 上看下来，需要补充两个点（后续需要支持）：
1、Y 轴可以从 0 开始。scale 直接设置的话，会导致负数情况的最小值也变成 0，因此可能需要一个 startOnZero 的参数，或者 statistic 扩展
2、轴标签允许 `formatter` 格式化

**Comment by pearmini** [comment]
> 从 PR 上看下来，需要补充两个点（后续需要支持）： 1、Y 轴可以从 0 开始。scale 直接设置的话，会导致负数情况的最小值也变成 0，因此可能需要一个 startOnZero 的参数，或者 statistic 扩展 2、轴标签允许 `formatter` 格式化

- 目前 Y 轴可以从 0 开发，设置 `scale : { zero: true }` 就可以了。
- formatter 这个后续会加上！

File: __tests__/unit/geometry/line.spec.ts [hunk 1]
```diff
@@ -0,0 +1,218 @@
...
+      ],
+      shapes: ['line', 'smooth'],
+    });
+  });
+
+  it('Line should draw basic line', () => {
+    const container = document.createElement('div');
+    const line = LineShape();
+    const shapes = plot({
+      mark: Line({}),
```
[the remark this answers is not in the document]
Comment (pearmini) [reply __tests__/unit/geometry/line.spec.ts:37]: annotation 和 geometry 都是 mark

File: src/infer/maybeSeries.ts [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+  const { color, series, ...rest } = encode;
+  if (series) return encode;
+  if (color !== undefined) return { color, series: color, ...rest };
+  return encode;
+}
+
+/**
+ * Produce series channel by color channel.
+ * It useful for line and area geometries to draw multiple series and areas.
+ * @example {x: [1, 2, 3]} -> {x: [[1], [2], [3]]}
```
[the remark this answers is not in the document]
Comment (pearmini) [reply src/infer/maybeSeries.ts:15]: 说明我是 copy 过来的，我改一下

File: src/runtime/plot.ts [hunk 3]
```diff
@@ -128,41 +130,55 @@ async function plotArea(options: G2Area, context: G2Context): Promise<void> {
...
+  const displays = normalizeDisplays(marks, components);
+  displays.sort((a, b) => a.zIndex - b.zIndex);
+
+  for (const { type, display } of displays) {
+    if (type === 'component') {
+      // Render components with corresponding bbox and scale(if required).
+      const scaleDescriptor = componentScale.get(display);
+      const scale = scaleDescriptor ? useScale(scaleDescriptor) : null;
+      const { field, domain } = scaleDescriptor;
+      const value = { field, domain };
```
[the remark this answers is not in the document]
Comment (pearmini) [reply src/runtime/plot.ts:153]: 之后可以看情况增加

File: src/scale/point.ts [hunk 4]
```diff
@@ -0,0 +1,12 @@
+import { Point as PointScale } from '@antv/scale';
+
+import { PointScale as PointScaleSpec } from '../spec';
```
[the remark this answers is not in the document]
Comment (pearmini) [reply src/scale/point.ts:3]: 我去改一下

**Review by visiky — APPROVED** [review]

**Pull request merged**